themes/default/content/blog/benchmarking-python-performance/index.md

```diff
@@ -0,0 +1,222 @@
+	name = f'pulumi-{str(i).rjust(3, "0")}'
+	aws.sns.Topic(name)
+```
+
+In this program, we can create 200 resources in parallel because none of them
+take inputs from other resources. This program should be entirely
+network-bound because Pulumi can issue all 200 API calls in parallel and wait 
+for AWS to provision the instances.
+[We discovered](https://github.com/pulumi/pulumi/issues/11116), however, 
+that it did not! Strangely, API calls were issued in an initial batch of 20; 
+as one completed, another would start.
+
+The culprit was the Python default future executor,
+[ThreadPoolExecutor](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ThreadPoolExecutor).
+We observed that benchmark was run on a four-core computer, and in Python 3.5
+to Python 3.7, the number of max workers is five times the number of cores, or 20.
+In (Python 3.8, this number was changed to `min(32, os.cpu_count() + 4)`.) We
+realized we shouldn't be using the default `ThreadExecutor`, and instead we
+should provide a `ThreadExecutor` with an adjusted number of `max_workers`
+based on the configured parallelism value. That way, when users run
+`pulumi up --parallel`, which issues an upper bound on parallel resource
+construction, the `ThreadExecutor` will respect that bound. We
+[merged a fix](https://github.com/pulumi/pulumi/pull/11122)
+that would plumb the value of `--parallel` through to a custom `ThreadExecutor`
+and measured the impact this change had on the performance of our benchmark.
+
+## Experimental Setup
+
+We designed and implemented two independent experiments to evaluate this change.
+The first experiment measures how well the patched Python runtime stacks up
+against the control group, Pulumi Python without the patch. The second experiment
+compares Pulumi Python to Pulumi TypeScript using the same benchmark ported
+to TypeScript. We used the awesome benchmarking tool
+[hyperfine](https://github.com/sharkdp/hyperfine) to record wall clock time as
+our indicator of performance.
+
+The experiments ran overnight on a 2021 MacBook Pro with 32GB RAM, the M1 chip,
+and 10 cores. Experimental code is
+[available on GitHub](https://github.com/RobbieMcKinstry/python-concurrency-experiments/tags),
+and release tags pin the version of the code used for each experiment.
+We also made an effort to run the experiments on a quiet machine connected
+to power. For all experiment groups, `--parallel` was unset, translating to
+unbounded parallelism.
+
+Before between samples, we ran `pulumi destroy –yes` to ensure a fresh 
+environment. Hyperfine measures shell startup time and subtracts the value
+before final measurements are recorded to more precisely represent the true
+cost of execution. All groups collected 20 samples each. We also discard 
+`stderr` and `stdout` to reduce noise associated with logging to a tty, but
+we do record the status code of each command so can show they executed successfully.
+
+## Python: Pre- and Post-patch
```

Comment:
Let me know if you think the data would be better visualized with a graph to show the difference, or if its fine as is.